syntax = "proto3";

package user.v1;

import "google/protobuf/timestamp.proto";

// User service for managing user accounts and authentication
service UserService {
    // Register a new user
    rpc RegisterUser(RegisterUserRequest) returns (UserProfileResponse) {}
    
    // Authenticate a user and return a token
    rpc AuthenticateUser(AuthenticateUserRequest) returns (AuthResponse) {}
    
    // Get the profile of the authenticated user
    rpc GetUserProfile(GetUserProfileRequest) returns (UserProfileResponse) {}
    
    // Update the profile of the authenticated user
    rpc UpdateUserProfile(UpdateUserProfileRequest) returns (UserProfileResponse) {}
}

// Request message for getting a user profile
message GetUserProfileRequest {
    string id = 1;
}

// Request message for registering a new user
message RegisterUserRequest {
    string username = 1;
    string password = 2;
    string email = 3;
}

// Request message for authenticating a user
message AuthenticateUserRequest {
    string username = 1;
    string password = 2;
}

// Response message for authentication
message AuthResponse {
    bool success = 1;
    string id = 2;
}

// Complete user profile information
message UserProfileResponse {
    string id = 1;
    string username = 2;
    string email = 3;
    string name = 4;
    string surname = 5;
    string date_of_birth = 6; // Format: YYYY-MM-DD
    string phone_number = 7;
    google.protobuf.Timestamp created_at = 8;
    google.protobuf.Timestamp updated_at = 9;
}

// Request message for updating a user profile
message UpdateUserProfileRequest {
    string name = 1;
    string surname = 2;
    string date_of_birth = 3; // Format: YYYY-MM-DD
    string email = 4;
    string phone_number = 5;
}

// Error response message
message ErrorResponse {
    int32 code = 1;
    string message = 2;
}
